feat: Add pulling repository

didrikmunther opened the pull request:
## Describe your changes

Fixes #6.

- Add library `jgit` (java git implementation)
- Add library `log4j` (required by `jgit`)
- Add library `commons-io` (to recursively delete directories)
- Implement an interface `PipelineComponent`. These are atomic components doing some action in the pipeline sequentially.
-  Implement `PipelinePuller` component, which pulls a repository from a push event.

## Checklist before requesting review

-   [x] Feature/fix PRs should add one atomic (as small as possible) feature or
        fix
-   [x] If your PR adds a feature or fix it also needs to add or modify at least
        one test
-   [x] All code in your PR needs to have been formatted
-   [x] The merge commit title should follow our commit prefix convention of
        either "feat", "fix", "chore", "doc", or "refactor"
-   [x] The merge commit body should reference at least one issue
-   [x] The code compiles and all the tests pass

didrikmunther committed:
Add pulling repository

didrikmunther committed:
Add test description for pipeline pull

didrikmunther committed:
Remove failing pipeline test code

didrikmunther committed:
Update pipeline pull test

didrikmunther committed:
Refactor push event test data

didrikmunther committed:
Fix pipeline test comment

didrikmunther committed:
Add proper javadoc for pull function

danielwis reviewed (approved):
LGTM. I like the idea of the directory structure you mention here (`pipelineDir/repository/commit_sha`), as this means that we have an easy way of storing the build logs for each build handled by the server.

psoder commented:
While the code in and of it self works fine I don't really like the way it's done.

IMHO functions should be as pure as possible and only use data passed to them through the function call. I.e. there shouldn't be any fields in the pipeline class  (i.e. `pipelineDir`, `event`, and `status`). In this case it more or less means making the functions `static` and removing the fields (yes there is a point in having a status field in order to be able to check the status for longer jobs, but that's a project for later).

The reason for this is that it makes the code more understandable since one only needs to be concerned with what goes into the function and nothing else. This is especially true in this case when the stages of the pipeline are, for all intents and purposes, independent of each other. Passing everything to the function also has the benefit of making testing and debugging much easier. 

I've taken the liberty of modifying this branch on https://github.com/soffan-group-25/ci-server/tree/f/add-pulling-repo2. I also added a config file for `log4j`. 

Good job on the code tho :)

edit:
I don't have a problem with the `Pipeline`-class having fields. It's more that since `pull`, `lint`, and `compile` are so separate from each other it's weird having them in the same class using the same fields.

didrikmunther committed:
Add pipeline puller class and respective tests

didrikmunther committed:
Remove commented out code

didrikmunther committed:
Add deleting testing repositories

didrikmunther commented:
@psoder you raise some very good points. I've modified the code so that the functions are more pure by the help of an abstracted layer `PipelineComponent`. I've also added tests for the git functionality.

psoder commented on app/src/main/java/ciserver/PipelinePuller.java:
```diff
@@ -0,0 +1,48 @@
+	 */
+	public PipelineStatus execute(String pipelineDir, PushEvent event) {
+		// Use the head_commit id as name for the repository directory
+		String directoryPath = String.format("%s/repositories/%s", pipelineDir, event.headCommit.id);
```
```suggestion
		String directoryPath = String.format("%s/%s/%s", pipelineDir, event.repository.name, event.headCommit.id);
```
It's probably good to also have the name of the repository in the path if you have multiple repositories.

didrikmunther replied:
You could, but the probability of hash collisions are so small, it shouldn't be a problem.

psoder commented on app/src/main/java/ciserver/Pipeline.java:
```diff
@@ -8,65 +11,64 @@ enum PipelineStatus {
-			return status;
+	private PipelineStatus _start() {
+		for (var component : components) {
+			var status = component.execute(pipelineDir, event);
+			if (status != PipelineStatus.Ok) {
+				return status;
+			}
 		}
 
 		return PipelineStatus.Ok;
 	}
 
 	/**
-	 * Start the pipeline.
+	 * Start the pipeline with the added components.
 	 * 
-	 * @return the status of the executed pipeline. OK if everything went ok.
+	 * @return the status of the pipeline execution
 	 */
 	public PipelineStatus start() {
 		this.status = PipelineStatus.InProgress;
 		this.status = _start();
 
 		return this.status;
 	}
```
What's the point of the `_start()` method? It seems like a very intentional decision, but I can't see the reason for it in this case. My suggestion would be to combine them into one.

```java
	/**
	 * Start the pipeline with the added components.
	 *
	 * @return the status of the pipeline execution
	 */
	public PipelineStatus start() {
		for (var component : components) {
			var status = component.execute(pipelineDir, event);
			if (status != PipelineStatus.Ok) {
				return status;
			}
		}

		return PipelineStatus.Ok;
	}
```

didrikmunther replied:
You're right, it was used before but not anymore.

psoder commented on app/src/main/java/ciserver/Pipeline.java:
```diff
@@ -8,65 +11,64 @@ enum PipelineStatus {
+	 * As one can individually test the pull, lint, compile functionality,
+	 * without having a bunch of public methods in the Pipeline class.
+	 */
+	ArrayList<PipelineComponent> components = new ArrayList<>();
```
```suggestion
	private ArrayList<PipelineComponent> components = new ArrayList<>();
```

didrikmunther replied:
Great!

psoder commented on app/src/main/java/ciserver/Pipeline.java:
```diff
@@ -8,65 +11,64 @@ enum PipelineStatus {
 class Pipeline {
-	String pipelineDir;
-	Commit commit;
+	final String pipelineDir;
+	final PushEvent event;
 	PipelineStatus status;
```
Fields should be private imo.

```suggestion

	private final String pipelineDir;
	private final PushEvent event;
```

didrikmunther replied:
I thought they were automatically, good catch!

psoder commented on app/src/test/java/ciserver/PipelinePullerTest.java:
```diff
@@ -0,0 +1,41 @@
+        assertEquals(status, PipelineStatus.Ok);
+
+        File directory = new File(
+                String.format("%s/repositories/%s", PipelineTest.PipelineTestingDirectory, event.headCommit.id));
```
See above.

```suggestion
                String.format("%s/%s/%s", PipelineTest.PipelineTestingDirectory, event.repository.name, event.headCommit.id));
```

psoder reviewed (changes requested):
Nicely done :+1:

Other than the changes mentioned I would probably rename `components` to `stages` in the `Pipeline` class.

psoder committed:
Incorporate changes

psoder commented:
In the interest of time I have committed the suggested changes to the branch. Feel free to revert them if they're no good. Also solves #45.

psoder commented:
On second thought I think it's probably better to do it like I've done it in #48. It's less abstraction and easier for the user to not fuck up when using it (i.e. there's no way for the user to compile without pulling). I still think it's good to have the different stages in separate files tho.

danielwis commented:
I think the way it's done in #48 has some good benefits, like having stages in a sequential order that _must_ be followed (as `pull()` will always execute before `lint()`, no matter the target etc.). However, having each stage as its own class implementing an `execute()` method makes for more flexibility as the pipeline grows larger.

For this case though, I think it may be better to stick with as little abstraction as possible such as in #48. This is easier to read and feels more scope-appropriate in terms of the assignment, as things are supposed to happen sequentially anyway. Example:
```java
Pipeline pipeline = new Pipeline(...);
pipeline.start(Target.COMPILE);
```
instead of
```java
// The stages below should be the same every time anyway
Pipeline pipeline = new Pipeline(...).withComponent(new PipelinePuller(), new PipelineLinter(), ...);
pipeline.start();
```

Regarding tests, these can be performed on `pull()`, `lint()` etc. instead as they are supposed to do the same thing as `execute()` anyway.

I think the abstraction is a nice touch, don't get me wrong, but even if we take that route I think a `target` is better than manually adding stages to the pipeline - at least in this assignment.

didrikmunther committed:
Remove abstraction of pipeline stages

didrikmunther committed:
Rename PipelineStage to TargetStage

psoder reviewed (approved):
Let's gamble, try merging
